Fails to compile under C++1y mode of GCC 4.9.2 with libstdc++ version 20150327 (libstdc++.so.6.0.20)

I create a new file `test.cpp` in the root directory of a fresh range-v3 clone that only contains the following:

~~~~~.cpp
#include <range/v3/core.hpp>

~~~~~

This compiles cleanly with any of:

    clang++ -std=c++11 -stdlib=libc++ -isystem include test.cpp -c
    clang++ -std=c++1y -stdlib=libc++ -isystem include test.cpp -c
    g++     -std=c++11                -isystem include test.cpp -c

...but fails with...

    g++     -std=c++1y                -isystem include test.cpp -c

~~~~~
In file included from include/range/v3/core.hpp:17:0,
                 from test.cpp:1:
include/range/v3/begin_end.hpp:40:24: error: ‘std::rbegin’ has not been declared
             using std::rbegin;
                        ^
include/range/v3/begin_end.hpp:41:24: error: ‘std::rend’ has not been declared
             using std::rend;
                        ^
In file included from include/range/v3/view_facade.hpp:24:0,
                 from include/range/v3/core.hpp:21,
                 from test.cpp:1:
include/range/v3/utility/basic_iterator.hpp:374:22: error: ‘constexpr const T& ranges::v3::basic_mixin<Cur>::get() const’ cannot be overloaded
             T const &get() const noexcept
                      ^
include/range/v3/utility/basic_iterator.hpp:368:16: error: with ‘constexpr T& ranges::v3::basic_mixin<Cur>::get() const’
             T &get() noexcept
                ^
include/range/v3/utility/basic_iterator.hpp:396:22: error: ‘constexpr const S& ranges::v3::basic_sentinel<S>::end() const’ cannot be overloaded
             S const &end() const noexcept
                      ^
include/range/v3/utility/basic_iterator.hpp:391:16: error: with ‘constexpr S& ranges::v3::basic_sentinel<S>::end() const’
             S &end() noexcept
                ^
include/range/v3/utility/basic_iterator.hpp:440:47: error: ‘constexpr const Cur& ranges::v3::basic_iterator<Cur, S>::pos() const’ cannot be overloaded
             RANGES_CXX14_CONSTEXPR Cur const &pos() const noexcept
                                               ^
include/range/v3/utility/basic_iterator.hpp:436:41: error: with ‘constexpr Cur& ranges::v3::basic_iterator<Cur, S>::pos() const’
             RANGES_CXX14_CONSTEXPR Cur &pos() noexcept
                                         ^
~~~~~

Context
-------

 * ***Compiler*** : GCC Ubuntu 4.9.2-10ubuntu13
 * ***Stdlib*** : GNU libstdc++ version 20150327 (/usr/lib/x86_64-linux-gnu/libstdc++.so.6.0.20)
 * ***OS*** : Ubuntu 15.04
 * ***Kernel*** : 3.19.0-25-generic #26-Ubuntu SMP Fri Jul 24 21:17:31 UTC 2015 x86_64

--- a/README.md
+++ b/README.md
@@ -26,7 +26,7 @@ Supported Compilers
 The code is known to work on the following compilers:
 
 - clang 3.4.0 (C++14 support requires at least clang 3.5.0)
-- GCC 4.9.0
+- GCC 4.9.0 (C++14 support requires GCC 5.2. Complex `constexpr` usage may be unreliable due to bugs in GCC, e.g. [PR67813](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=67813), as witnessed by the failure of the constexpr-core test.)
 
 **Development Status:** This code is fairly stable, well-tested, and suitable for casual use, although currently lacking documentation. No promise is made about support or long-term stability. This code *will* evolve without regard to backwards compatibility.
 

--- a/include/meta/meta.hpp
+++ b/include/meta/meta.hpp
@@ -18,6 +18,7 @@
 #include <cstddef>
 #include <initializer_list>
 #include <type_traits>
+#include <utility>
 #include <meta/meta_fwd.hpp>
 
 #if defined(__clang__)
@@ -190,12 +191,12 @@ namespace meta
         /// An integral constant wrapper around the result of incrementing the wrapped integer \c
         /// T::type::value.
         template <typename T>
-        using inc = std::integral_constant<decltype(T::type::value), T::type::value + 1>;
+        using inc = std::integral_constant<decltype(T::type::value + 1), T::type::value + 1>;
 
         /// An integral constant wrapper around the result of decrementing the wrapped integer \c
         /// T::type::value.
         template <typename T>
-        using dec = std::integral_constant<decltype(T::type::value), T::type::value - 1>;
+        using dec = std::integral_constant<decltype(T::type::value - 1), T::type::value - 1>;
 
         /// An integral constant wrapper around the result of adding the two wrapped integers
         /// \c T::type::value and \c U::type::value.
@@ -215,7 +216,7 @@ namespace meta
         /// \c T::type::value and \c U::type::value.
         /// \ingroup math
         template <typename T, typename U>
-        using multiplies = std::integral_constant<decltype(T::type::value *U::type::value),
+        using multiplies = std::integral_constant<decltype(T::type::value * U::type::value),
                                                   T::type::value * U::type::value>;
 
         /// An integral constant wrapper around the result of dividing the two wrapped integers \c
@@ -225,8 +226,8 @@ namespace meta
         using divides = std::integral_constant<decltype(T::type::value / U::type::value),
                                                T::type::value / U::type::value>;
 
-        /// An integral constant wrapper around the remainder of dividing the two wrapped integers
-        /// \c T::type::value and \c U::type::value.
+        /// An integral constant wrapper around the result of negating the wrapped integer
+        /// \c T::type::value.
         /// \ingroup math
         template <typename T>
         using negate = std::integral_constant<decltype(-T::type::value), -T::type::value>;
@@ -278,7 +279,7 @@ namespace meta
         /// integers \c T::type::value and \c U::type::value.
         /// \ingroup math
         template <typename T, typename U>
-        using bit_and = std::integral_constant<decltype(T::type::value &U::type::value),
+        using bit_and = std::integral_constant<decltype(T::type::value & U::type::value),
                                                T::type::value & U::type::value>;
 
         /// An integral constant wrapper around the result of bitwise-or'ing the two wrapped
@@ -295,7 +296,7 @@ namespace meta
         using bit_xor = std::integral_constant<decltype(T::type::value ^ U::type::value),
                                                T::type::value ^ U::type::value>;
 
-        /// An integral constant wrapper around the result of bitwise-complimenting the wrapped
+        /// An integral constant wrapper around the result of bitwise-complementing the wrapped
         /// integer \c T::type::value.
         /// \ingroup math
         template <typename T>
@@ -894,9 +895,17 @@ namespace meta
 
         /// Logically and together all the Boolean parameters
         /// \ingroup logical
+#if (__GNUC__ == 5) && (__GNUC_MINOR__ == 1) && !defined(__clang__)
+        // Alternative formulation of and_c to workaround
+        // https://gcc.gnu.org/bugzilla/show_bug.cgi?id=66405
+        template <bool... Bools>
+        using and_c = std::is_same<integer_sequence<bool, true, Bools...>,
+                                   integer_sequence<bool, Bools..., true>>;
+#else
         template <bool... Bools>
         using and_c = std::is_same<integer_sequence<bool, Bools...>,
                                    integer_sequence<bool, (Bools || true)...>>;
+#endif
 
         /// Logically and together all the integral constant-wrapped Boolean parameters, \e without
         /// doing short-circuiting.
@@ -1284,6 +1293,13 @@ namespace meta
 
                 template <typename... Ts>
                 static id<list<Ts...>> eval(VoidPtrs..., id<Ts> *...);
+            };
+
+            template <>
+            struct drop_impl_<list<>>
+            {
+                template <typename...Ts>
+                static id<list<Ts...>> eval(id<Ts> *...);
             };
 
             template <typename List, typename N>
@@ -2804,6 +2820,7 @@ namespace meta
         ///////////////////////////////////////////////////////////////////////////////////////////////
         // integer_sequence
 
+#ifndef __cpp_lib_integer_sequence
         /// A container for a sequence of compile-time integer constants.
         /// \ingroup integral
         template <typename T, T... Is>
@@ -2813,6 +2830,7 @@ namespace meta
             /// \return `sizeof...(Is)`
             static constexpr std::size_t size() noexcept { return sizeof...(Is); }
         };
+#endif
 
         /// \cond
         namespace detail

--- a/include/meta/meta_fwd.hpp
+++ b/include/meta/meta_fwd.hpp
@@ -15,6 +15,8 @@
 #ifndef META_FWD_HPP
 #define META_FWD_HPP
 
+#include <utility>
+
 #ifndef META_DISABLE_DEPRECATED_WARNINGS
 #ifdef __cpp_attribute_deprecated
 #define META_DEPRECATED(MSG) [[deprecated(MSG)]]
@@ -35,8 +37,12 @@ namespace meta
 {
     inline namespace v1
     {
+#ifdef __cpp_lib_integer_sequence
+        using std::integer_sequence;
+#else
         template <typename T, T...>
         struct integer_sequence;
+#endif
 
         template <typename... Ts>
         struct list;

--- a/test/CMakeLists.txt
+++ b/test/CMakeLists.txt
@@ -1,9 +1,9 @@
 # Strange, need to use static linking here or tests fail.
-if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" )
-  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static")
-endif()
+#if("${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU" )
+#  set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -static")
+#endif()
 
-set(CMAKE_CXX_COMPILER_FLAGS "${CMAKE_CXX_COMPILER_FLAGS} -g")
+#set(CMAKE_CXX_COMPILER_FLAGS "${CMAKE_CXX_COMPILER_FLAGS} -g")
 
 add_subdirectory(action)
 add_subdirectory(algorithm)

--- a/test/constexpr_core.cpp
+++ b/test/constexpr_core.cpp
@@ -181,10 +181,13 @@ RANGES_CXX14_CONSTEXPR auto test_array() -> bool {
 
     auto beg = ranges::begin(a);
     auto three = ranges::next(beg, 2);
-    // ranges::iter_swap(beg, three);
-    // if (*beg != 3) { return false; }
-    // if (*three != 1) { return false; }
-    // ranges::iter_swap(beg, three);
+
+    if ((false)) {
+      ranges::iter_swap(beg, three);
+      if (*beg != 3) { return false; }
+      if (*three != 1) { return false; }
+      ranges::iter_swap(beg, three);
+    }
 
     if (!test_its(a)) { return false; }
     if (!test_cits(a)) { return false; }

